src/onyx/api.clj

```diff
@@ -167,34 +165,61 @@
+   :job-id id})
+
 (defn ^{:added "0.6.0"} submit-job
   "Takes a peer configuration, job map, and optional monitoring config,
-   sending the job to the cluster for eventual execution."
+   sending the job to the cluster for eventual execution. Returns a map
+   with :success? indicating if the job was submitted to ZooKeeper. The job map
+   may contain a :metadata key, among other keys described in the user
+   guide. The :metadata key may optionally supply a :job-id value. Repeated
+   submissions of a job with the same :job-id will be treated as an idempotent
+   action. If a job has been submitted more than once, the original task IDs
+   associated with the catalog will be returned, and the job will not run again,
+   even if it has been killed or completed. It is undefined behavior to submit
+   two jobs with the same :job-id metadata whose :workflow, :catalog,
+   :flow-conditions,etc are not equal."
   ([peer-client-config job]
    (submit-job peer-client-config job {:monitoring :no-op}))
   ([peer-client-config job monitoring-config]
    (let [result (validate-submission job peer-client-config)]
      (if (:success? result)
-       (let [id (java.util.UUID/randomUUID)
+       (let [job (update-in job [:metadata :job-id] #(or % (UUID/randomUUID)))
+             job-hash (hash-job job)
+             id (get-in job [:metadata :job-id])
              tasks (planning/discover-tasks (:catalog job) (:workflow job))
              entry (create-submit-job-entry id peer-client-config job tasks)
-             client (component/start (system/onyx-client peer-client-config monitoring-config))]
-         (extensions/write-chunk (:log client) :catalog (:catalog job) id)
-         (extensions/write-chunk (:log client) :workflow (:workflow job) id)
-         (extensions/write-chunk (:log client) :flow-conditions (:flow-conditions job) id)
-         (extensions/write-chunk (:log client) :lifecycles (:lifecycles job) id)
-         (extensions/write-chunk (:log client) :windows (:windows job) id)
-         (extensions/write-chunk (:log client) :triggers (:triggers job) id)
-         (extensions/write-chunk (:log client) :job-metadata (:metadata job) id)
-
-         (doseq [task tasks]
-           (extensions/write-chunk (:log client) :task task id))
-
-         (extensions/write-log-entry (:log client) entry)
-         (component/stop client)
-         {:success? true
-          :job-id id
-          :task-ids (zipmap (map :name tasks) tasks)})
+             client (component/start (system/onyx-client peer-client-config monitoring-config))
+             status (extensions/write-chunk (:log client) :job-hash job-hash id)]
+         (if status
+           (serialize-job-to-zookeeper client id job tasks entry)
+           (let [written-hash (extensions/read-chunk (:log client) :job-hash id)]
+             (if (= written-hash job-hash)
+               (serialize-job-to-zookeeper client id job tasks entry)
+               {:success? false}))))
```

Comment:
I think all of the submitters should still go through the process of writing all of the job chunks, in case one dies midway. Otherwise we don't know if the submission was successful.

We no longer need to check the return codes of any of the writes. As long as one won, all is well.